td/tracking/views.py

```diff
@@ -95,10 +98,193 @@ class CharterUpdate(LoginRequiredMixin, UpdateView):
     form_class = CharterForm
     template_name_suffix = "_update_form"
 
+    # Overwritten to redirect upon valid submission
     def form_valid(self, form):
         self.object = form.save()
         messages.info(self.request, "Project charter has been updated")
-        return redirect("tracking:charter_add_success", pk=self.object.id)
+        return redirect("tracking:charter_add_success", obj_type="charter", pk=self.object.id)
+
+
+# -------------------------------------------------------------------
+# NOTE: Charter detail will be integrated to the language detail page
+# -------------------------------------------------------------------
+def charter(request, pk):
+    charter = get_object_or_404(Charter, pk=pk)
+    messages.info(request, "This page provides a link to edit a charter, but is still being worked on.")
+    context = {
+        "charter": charter,
+    }
+
+    return render(request, "tracking/charter_detail.html", context)
+
+
+# -------------------------------- #
+#            EVENT VIEWS           #
+# -------------------------------- #
+
+
+class EventAddView(LoginRequiredMixin, CreateView):
+    model = Event
+    form_class = EventForm
+
+    # Overwritten to include initial values
+    def get_initial(self):
+        return {
+            "start_date": timezone.now(),
+            "created_by": self.request.user.username,
+        }
+
+    # Overwritten to pass URL argument to forms.py
+    def get_form_kwargs(self, **kwargs):
+        keywords = super(EventAddView, self).get_form_kwargs(**kwargs)
+        if "pk" in self.kwargs:
+            keywords["pk"] = self.kwargs["pk"]
+        return keywords
+
+    # Overwritten to include custom data
+    def get_context_data(self, *args, **kwargs):
+        context = super(EventAddView, self).get_context_data(**kwargs)
+        context["translators"] = self.get_translator_data(self)
+        context["facilitators"] = self.get_facilitator_data(self)
+        context["materials"] = self.get_material_data(self)
+        return context
+
+    # Overwritten to execute custom save and redirect upon valid submission
+    def form_valid(self, form):
+        self.object = form.save()
```

Comment:
make this:

`event = self.object = form.save()`

Then reuse `event` object in every other block in this method and avoid the lookup.

Everywhere you get a list of ids and then loop over them, looking up the object and adding to the m2m manager, get a list of them and expand into the `m2m.add` method:

e.g.:
`event.translators.add(*list(Translator.objects.filter(id__in=translator_ids)))`